tests/codex/utils/testasynciter.nim

```diff
@@ -0,0 +1,130 @@
+import std/sugar
+
+import pkg/questionable
+import pkg/chronos
+import pkg/codex/utils/asynciter
+
+import ../../asynctest
+import ../helpers
+
+asyncchecksuite "Test AsyncIter":
+
+  test "Should be finished":
+    let iter = AsyncIter[int].empty()
+
+    check:
+      iter.finished == true
+
+  test "Should map each item using `map`":
+    let
+      iter1 = AsyncIter[int].new(0..<5).delayBy(10.millis)
```

Comment:
Interesting, why the `delayBy()` part there?

Reply:
Just to make sure that we are doing proper async flow, so yielding control back to chronos scheduler for the time of the sleep. Without this it would be synchronous calls only and essentially the same as non-async iter.

Reply:
Hmm, I guess it makes sense, although I don't think it is necessary? Because every time you call `await`, you are yielding control, and there are `await` calls in the `flatMap` and then in collecting, so you should be good in the "yielding control" part. In the same time, I guess it does not really hurt anything to do it like this.

Reply:
I might be wrong, but my impression was that Chronos will not yield control back to the scheduler if the future that we want to await for is already completed, which I think is the case for simple proc with `{.async.}` pragma returning a value immediately (without any await for IO or sleep).

Reply:
Yep, I didn't follow the whole flow but you need to await on a non-completed Future [otherwise Chronos will not add a callback](https://github.com/codex-storage/nim-chronos/blob/master/chronos/internal/asyncfutures.nim#L379); i.e., it will run your code to completion.

Reply:
Interesting! TIL - today I learned ;-)